[resolver] use webpack-resolver for core logic

spalger opened the pull request:
This PR removes the custom resolver logic and instead wraps the `eslint-import-resolver-webpack` module, feeding it a custom configuration so that the resolver behaves like Kibana's internal version of webpack.

This also adds three settings:
- `rootPackageName`: the name of the package that should be considered the "root" package, useful for packages like Kibana which contain multiple packages so that a file like `src/core_plugins/console/index.js` resolves as a child of kibana rather than the console plugin package.
- `pluginPaths`: an array of paths to plugin folders
- `pluginDirs`: an array of paths (relative to the project root) that contain multiple plugin folders

spalger committed:
[resolver] use webpack-resolver for core logic

spalger committed:
use better name for `fromRoot()` function

spalger committed:
[resolver] add default aliases from test_bundle

spalger committed:
remove trailing whitespace

spalger committed:
enabled node eslint-env

w33ble commented on index.js:
```diff
@@ -1,214 +1,106 @@
+  pluginPaths: [],
+  pluginDirs: [
+    '../kibana/plugins',
+    '../kibana/core_plugins',
```
This path is wrong, it's `kibana/src/core_plugins`

w33ble committed:
always look in kibana plugins

also observe the kibanaPath

w33ble committed:
add eslint-plugin-import

w33ble committed:
replace find-root with pkg-up

w33ble committed:
add options info to readme

w33ble committed:
use pluginName config, if provided

recurse up paths until a package.json file with a matching name is found, if pluginName is provided. throw if none is found

w33ble committed:
break index into multiple module files

w33ble committed:
add debugging info to readme

w33ble committed:
switch pluginName to rootPackageName

w33ble committed:
switch pluginName to rootPackageName

spalger commented on index.js:
```diff
@@ -1,213 +1,130 @@
-  const matches = glob.sync(globPattern, globOptions);
-  debug(`checking in ${checkPath}, matched ${matches.length}`);
+  const getRootPackageDir = (dir) => {
+    const pkgFile = pkgUp.sync(dir);
```
Do we really need `pkgUp` for this? We're basically already doing what it is doing in this function.

spalger replied:
Start at the path and recursively look for package.json files that match the `rootPackageName` or just exist if there isn't a `rootPackageName`.

w33ble replied:
I suppose you're right, we kinda don't...

spalger commented on index.js:
```diff
@@ -1,213 +1,130 @@
  * Resolves the path to Kibana, either from default setting or config
  */
-function getKibanaPath(config, file, rootPath) {
+function getKibanaPath(config, rootPath) {
```
Can we rename `rootPath` to `projectRoot`?

w33ble committed:
Merge branch 'modularize' into use-webpack-resolver

w33ble committed:
rename rootPath to projectRoot

w33ble committed:
remove pkg-up dependency

w33ble committed:
move resolution caching back to index
